### ISSUE TITLE
Implement NumPy sorting routines

### ISSUE THREAD

**Issue by shoyer** [google/jax#2079]
xref https://docs.scipy.org/doc/numpy/reference/routines.sort.html

Extracted from #70:

 - [x] np.sort
 - [x] np.lexsort
 - [x] np.argsort
 - [x] np.msort
 - [x] np.sort_complex
 - [x] np.partition
 - [ ] np.argpartition

All of these should be relatively doable, though it's worth noting that I don't think XLA currently has efficient partitioning schemes.

**Comment by Vaibhavs10** [comment]
@shoyer and @srvasude - Can I help any of these features into JAX?

P.S. I'm a beginner in terms of Open Source Contributions, so checking with you both before getting my hands dirty with this.

**Comment by jakevdp** [comment]
Hi @Vaibhavs10 - this is a bit complicated, because sorts must be implemented in terms of [XLA operations](https://www.tensorflow.org/xla/operation_semantics#sort), which are exposed in JAX in https://github.com/google/jax/blob/master/jax/lax/lax.py.

Of the ones listed above, I think `lex_sort` and `sort_complex` would be doable using XLA's sorting of multiple inputs with a custom comparator, but unfortunately `lax.sort` doesn't yet support passing a custom comparator: that would have to be added as a precursor to this work.

**Comment by jakevdp** [comment]
Actually, it looks like `np.sort_complex` would be a fairly good place to start. `lax.sort()` already handles complex input correctly as far as I can tell. What would need to happen is to define `jax.numpy.sort_complex` similarly to `sort` here: https://github.com/google/jax/blob/c1aeb8b3fe94f2a064ead43d8d9a1c35982d11d5/jax/numpy/lax_numpy.py#L3141-L3151

with modifications that make it follow the semantics of `np.sort_complex` (things like: real inputs result in complex outputs, etc.

And then add a test of `sort_complex` similar to the test of `sort` here: https://github.com/google/jax/blob/c1aeb8b3fe94f2a064ead43d8d9a1c35982d11d5/tests/lax_numpy_test.py#L2569-L2589

If you want to try this out, that would be great! If you need, feel free to open a draft pull request and tag me if you have questions.

**Comment by Vaibhavs10** [comment]
Thanks @jakevdp I’ll look at this tomorrow in my day time and raise a PR hopefully by weekend.

**Comment by tudorcebere** [comment]
Is this issue still open regarding partitioning schemes?

**Comment by jakevdp** [comment]
Yes, I don't think anyone has worked on partitions yet. It's not trivial to do, because they're not (yet?) implemented in XLA

It may be possible to implement using scans (similar to how `jnp.searchsorted` is implemented), but in the end it might actually be faster to implement partitioning via a full sort, since XLA knows how to do this efficiently.

**Comment by milind-soni** [comment]
> Yes, I don't think anyone has worked on partitions yet. It's not trivial to do, because they're not (yet?) implemented in XLA
> 
> It may be possible to implement using scans (similar to how `jnp.searchsorted` is implemented), but in the end it might actually be faster to implement partitioning via a full sort, since XLA knows how to do this efficiently.

can you please tell me how to get started with implementing this. I am a beginner to open source.

**Comment by jakevdp** [comment]
Hi - thanks for the interest! Are you asking how to get started with open source github contributions in general, or how to get started with JAX in particular?

**Comment by milind-soni** [comment]
> Hi - thanks for the interest! Are you asking how to get started with open source github contributions in general, or how to get started with JAX in particular?

Thanks for the quick reply. I want to get started with implementing the partition and argpartition functions.

**Comment by jakevdp** [comment]
OK - as mentioned above, I don't know if there's really any good way to do that at the moment. XLA has no partial sorts, so the only option would be to implement e.g. a partial [quicksort](https://en.wikipedia.org/wiki/Quicksort) from scratch using [control flow operators](https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators). A similar approach for implementing binary search was used for `jnp.searchsorted`, which you can see [here](https://github.com/google/jax/blob/main/jax/_src/numpy/lax_numpy.py#L5746-L5758), although sorting is somewhat more complicated. Since lax control flow operations tend to be fairly slow, I'm not sure if it would be worth going down that route; it may be better to just wait and hope that XLA implements partial sorts at some point in the future.

**Comment by milind-soni** [comment]
> OK - as mentioned above, I don't know if there's really any good way to do that at the moment. XLA has no partial sorts, so the only option would be to implement e.g. a partial [quicksort](https://en.wikipedia.org/wiki/Quicksort) from scratch using [control flow operators](https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators). A similar approach for implementing binary search was used for `jnp.searchsorted`, which you can see [here](https://github.com/google/jax/blob/main/jax/_src/numpy/lax_numpy.py#L5746-L5758), although sorting is somewhat more complicated. Since lax control flow operations tend to be fairly slow, I'm not sure if it would be worth going down that route; it may be better to just wait and hope that XLA implements partial sorts at some point in the future.

Got it. would love to have a shot at some other beginner issue. Can you suggest something?

**Comment by jakevdp** [comment]
It looks like the number of [good first issue](https://github.com/google/jax/labels/good%20first%20issue) issues has dwindled recently... there may be a few routines in #70 that would be a good. `np.insert` could be an interesting one, also `np.polyfit` and `np.polydiv`

**Comment by samuela** [comment]
Just want to mention that `argpartition` is especially useful when it comes to implementing top-N accuracy metrics.

**Comment by jakevdp** [comment]
For that application, you might check out [`lax.top_k`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html) and [`lax.approx_max_k`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.approx_max_k.html?highlight=approx_top_k).

**Comment by samuela** [comment]
oooh thank you @jakevdp!!
